### PR TITLE
Handle race conditions in AsyncQueryBatchCursor.

### PR DESCRIPTION
**Pull request by rozza**
Handle the scenario where the cursor is closed and there is a getMore call inflight.

JAVA-2445

### CODE REVIEW

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 1]
```diff
@@ -243,17 +284,24 @@ private void handleGetMoreQueryResult(final AsyncConnection connection, final Si
...
             } else {
-                QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(result.getDocument("cursor"),
-                                                                                connectionSource.getServerDescription().getAddress());
-                handleGetMoreQueryResult(connection, callback, queryResult);
+
+                if (isClosed() || localConnectionSource == null) {
+                    callback.onResult(null, GET_MORE_AFTER_CLOSE_EXCEPTION);
+                } else {
+                    QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(result.getDocument("cursor"),
+                            localConnectionSource.getServerDescription().getAddress());
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: The locking is not removing all the race conditions.   If close() is called at just the wrong time, localConnectionSource.release() could be called between the time we get our method-local reference to it and the time we try to use it, and we'd end up trying to use an instance with a zero reference count, which will thrown an exception.

What if instead of all the locking we just get the ServerAddress from:

     connection.getDescription().getServerAddress()

We know that's good because we've retained a reference to it and haven't yet released it.

**Comment by rozza** [comment]
Updated

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 2]
```diff
@@ -89,8 +91,15 @@
...
+            if (!closed) {
+                closed = true;
+                callKillCursor = true;
+            }
+        }
+
+        if (callKillCursor) {
             killCursor(null);
         }
     }
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Done

**Comment by rozza** [comment]
@jyemin updated and separated out the case where there is no AsyncConnectionSource

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 3]
```diff
@@ -130,41 +120,56 @@ public boolean isClosed() {
...
+        return oldCursor;
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void getMore(final SingleResultCallback<List<T>> callback) {
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
         connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: There's still a race between next and killCursorOnClose.  Though unlikely, it could end up in a situation where there is a non-null cursor but the connectionSource's reference count is 0, and getConnection throws an exception as a result.

One way to fix this is for next() to atomically get the cursor and also retain a reference to the connectionSource, and for killCursorOnClose() to atomically get the cursor using the same lock.  The latter doesn't need to retain the connectionSource though, because it's the only one that would release it back to 0.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 4]
```diff
@@ -50,18 +51,14 @@
     private final int limit;
     private final Decoder<T> decoder;
     private final long maxTimeMS;
-    private volatile AsyncConnectionSource connectionSource;
+    private final Object lock = new Object();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Though see below, as there still may be a need for a lock to coordinate access to the connectionSource and the cursor.

File: driver-core/src/main/com/mongodb/operation/AsyncSingleBatchQueryCursor.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import com.mongodb.connection.QueryResult;
+
+import java.util.List;
+
+class AsyncSingleBatchQueryCursor<T> implements AsyncBatchCursor<T> {
+    private volatile QueryResult<T> firstBatch;
+    private volatile boolean closed;
+
+    AsyncSingleBatchQueryCursor(final QueryResult<T> firstBatch) {
+        this.firstBatch = firstBatch;
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncSingleBatchQueryCursor.java]: 👍

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 4 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: 👍

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 3 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: I'm still unclear how `connectionSource` can have a reference count of zero.  We retain it in the constructor and it is released via `killCursorOnClose()`.  If the limit is reached, then we call `killCursor()`.  Here the connection source is retained there to ensure `close()` manages the final reference.

However, I now changed it so `next()` sets the cursor to null, now only the getMore can set the cursor.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 6]
```diff
@@ -177,48 +182,55 @@ private BsonDocument asGetMoreCommandDocument() {
...
+            });
+        } else {
             connectionSource.release();
-            connectionSource = null;
         }
     }
 
-    private void killCursorAsynchronouslyAndReleaseConnectionAndSource(final AsyncConnection connection, final ServerCursor localCursor,
-                                                                       final AsyncConnectionSource localConnectionSource) {
+    private void killCursor(final AsyncConnection connection) {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Good Catch

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 3 above)
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: The race I was referring to is between next() and close():

T1: next() gets non-null cursor
T2: close() sets cursor to null
T3: close() kills the cursor
T4: close() (asynchronously) releases connectionSource
T5: next() calls connectionSource.getConnection, but refcount is 0

The most recent commit has a similar race condition.

T1: next() sets cursor to null, and gets a non-null cursor
T2: close() sets cursor to null, and gets a null cursor
T3: close() (synchronously) releases connectionSource since cursor is null
T4: next() calls connectionSource.getConnection, but refcount is 0

Is also changes the behavior of the cursor, as now there is no way to kill a cursor while a getMore is in flight.  And once the getMore completes, the cursor won't be closed once it's done, potentially leaving an open cursor on the server.  I don't think we should make this change.

Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: > The most recent commit has a similar race condition.
> 
> T1: next() sets cursor to null, and gets a non-null cursor
> T2: close() sets cursor to null, and gets a null cursor
> T3: close() (synchronously) releases connectionSource since cursor is null
> T4: next() calls connectionSource.getConnection, but refcount is 0

T4  would throw an error as its called after close(), but you're right we don't close the final getMore cursor.

I've added a new patch and some more unit tests.

File: driver-core/src/test/functional/com/mongodb/ClusterFixture.java [hunk 7]
```diff
@@ -79,7 +79,7 @@
     public static final String DEFAULT_URI = "mongodb://localhost:27017";
     public static final String MONGODB_URI_SYSTEM_PROPERTY_NAME = "org.mongodb.test.uri";
     private static final String DEFAULT_DATABASE_NAME = "JavaDriverTest";
-    public static final long TIMEOUT = 60L;
+    public static final long TIMEOUT = 1L;
```
Comment (jyemin) [inline driver-core/src/test/functional/com/mongodb/ClusterFixture.java]: This seems short.  Why the change?

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 8]
```diff
@@ -177,53 +168,54 @@ private BsonDocument asGetMoreCommandDocument() {
         return document;
     }
 
+    private void killCursorOnClose() {
+        final ServerCursor localCursor = getCursorForKillCursorOnClose();
+        if (localCursor != null) {
+            connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
+                @Override
+                public void onResult(final AsyncConnection connection, final Throwable t) {
+                    killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, localCursor);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Check for t != null.  Otherwise connection could be null and we'll get an NPE

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 9]
```diff
@@ -271,11 +267,36 @@ public void onResult(final BsonDocument result, final Throwable t) {
...
                 handleGetMoreQueryResult(connection, callback, result);
             }
         }
     }
+
+    private ServerCursor getCursorForNext() {
+        ServerCursor localCursor;
+        synchronized (lock) {
+            localCursor = cursor.get();
+            if (localCursor == null) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Is it necessary to close the cursor at this point?  Can this be simplified to:

     if (localCursor != null) {
           connectionSource.retain();
     }

I think so since the next method already calls close() if this method returns a null cursor.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 10]
```diff
@@ -71,100 +70,92 @@
...
-        }
+        this.cursor = new AtomicReference<ServerCursor>(firstBatch.getCursor());
+        this.connectionSource = notNull("connectionSource", connectionSource);
         this.count += firstBatch.getResults().size();
-        if (limitReached()) {
-             killCursor(connection);
+
+        connectionSource.retain();
+        if (firstBatch.getCursor() != null && limitReached()) {
+            killCursor(connection);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: There's a bug here, and I see now that this is an untested code path at least in the functional tests.

The problem is that killCursor now releases the connection, but since this connection is not retained here the refcount is going to drop below 0 once the caller releases it.  So there has to be an extra retain somewhere, either here or in killCursor.   Note that the killCursor method currently retains the connection, but that retain is no longer there.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 11]
```diff
@@ -177,53 +168,54 @@ private BsonDocument asGetMoreCommandDocument() {
...
         });
     }
 
     private void handleGetMoreQueryResult(final AsyncConnection connection, final SingleResultCallback<List<T>> callback,
                                           final QueryResult<T> result) {
-        if (result.getResults().isEmpty() && result.getCursor() != null) {
-            getMore(connection, callback);
+        cursor.getAndSet(result.getCursor());
+        if (isClosed() && result.getCursor() != null) {
+            killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, result.getCursor());
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: if isClosed() already why is it our responsibility to kill the cursor?  Shouldn't close() have done that for us?

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java [hunk 12]
```diff
@@ -177,53 +168,54 @@ private BsonDocument asGetMoreCommandDocument() {
         return document;
     }
 
+    private void killCursorOnClose() {
+        final ServerCursor localCursor = getCursorForKillCursorOnClose();
+        if (localCursor != null) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: if localCursor == null the connectionSource is never released.  There seem to be tests for this code path but no assertions about the connectionSource reference count.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 10 above)
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: It's hit in the unit tests but the mocks don't check the reference counts.

**Comment by rozza** [comment]
I'm going to have to admint defeat and throw in the towel on this. Its a waste of our both our time.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 11 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: Because the cursor now is not null but close may have been called. This is the case you mentioned in the previous review that needed fixing.

File: driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java (hunk 9 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java]: It maybe the current tests have been historically wrong but simplifying this means some tests will fail as the connectionSource wont be released.

**Comment by rozza** [comment]
I'm going to start afresh and will open a new PR with the results.
